test/hotspot/jtreg/compiler/c2/irTests/TestCountedLoopCastIV.java

```diff
@@ -0,0 +1,176 @@
+    // which will create CastIINodes for loop induction variable.
+    // In this case, the loop start, limit and stride are
+    // all constants.
+    @Test
+    @IR(counts = {IRNode.COUNTED_LOOP, "1" })
+    static void test1() {
+        for (int i = 0; i < LEN; i += 16) {
+            Objects.checkIndex(i, LEN - 3);
+            int a = in[i + 3];
+            Objects.checkIndex(i, LEN - 15);
+            out[i + 15] = a;
+        }
+    }
+
+    @Run(test = "test1")
+    public static void runTest1() {
+        test1();
+        verify(in, out, 0, LEN, 16, 3, 15);
+    }
+
+    // Similar to test1, but the loop limit is a variable.
+    @Test
+    @IR(counts = {IRNode.COUNTED_LOOP, "1" })
+    static void test2(int limit) {
+        for (int i = 0; i < limit; i += 16) {
+            Objects.checkIndex(i, LEN - 3);
+            int a = in[i + 3];
+            Objects.checkIndex(i, LEN - 15);
+            out[i + 15] = a;
+        }
+    }
+
+    @Run(test = "test2")
+    private void runTest2() {
+        cleanUp();
+        test2(100);
+        verify(in, out, 0, 100, 16, 3, 15);
+
+        cleanUp();
+        test2(500);
+        verify(in, out, 0, 500, 16, 3, 15);
+
+        cleanUp();
+        test2(LEN);
+        verify(in, out, 0, LEN, 16, 3, 15);
+    }
+
+    // Similar to test1 and test2, but the loop is a
+    // while loop with a variable start and limit.
+    @Test
+    @IR(counts = {IRNode.COUNTED_LOOP, "1" })
+    static void test3(int start, int limit) {
+        int i = start;
+        while (i < limit) {
+            Objects.checkIndex(i, LEN);
+            int a = in[i];
+            Objects.checkIndex(i, LEN - 3);
+            out[i + 3] = a;
+            i++;
+        }
+    }
+
+    @Run(test = "test3")
+    private void runTest3() {
+        cleanUp();
+        test3(0, 100);
+        verify(in, out, 0, 100, 1, 0, 3);
+
+        cleanUp();
+        test3(128, 500);
+        verify(in, out, 128, 500, 1, 0, 3);
+
+        cleanUp();
+        test3(LEN - 128, LEN - 3);
+        verify(in, out, LEN - 128, LEN - 3, 1, 0, 3);
+    }
+
+    // Similar to test3, but the type of induction variable
+    // is long.
+    @Test
+    @IR(counts = {IRNode.COUNTED_LOOP, "1" })
+    static void test4(long start, long limit) {
+        for (long i = start; i < limit; i++) {
+            Objects.checkIndex(i, LEN);
+            int a = in[(int) i];
+            Objects.checkIndex(i, LEN - 3);
+            out[(int) i + 3] = a;
+        }
+    }
+
+    @Run(test = "test4")
+    private void runTest4() {
+        cleanUp();
+        test3(0, 100);
+        verify(in, out, 0, 100, 1, 0, 3);
+
+        cleanUp();
+        test3(128, 500);
+        verify(in, out, 128, 500, 1, 0, 3);
+
+        cleanUp();
+        test3(LEN - 128, LEN - 3);
+        verify(in, out, LEN - 128, LEN - 3, 1, 0, 3);
+    }
+
+    public static void main(String[] args) {
+        TestFramework.runWithFlags("-XX:LoopUnrollLimit=0");
```

Comment:
What is the reason for the flag here? Do you really need it?

Reply:
Thanks so much for your review! This flag prevents the loop being unrolled and splited (i.e. pre-main-post loop mode) as well. So that the compiler can get just one `CountedLoop` in the case, and we can make sure it is generated by the loop exactly. Checking the count of `CountedLoop` >0 is also fine to me without this flag. I just want to avoid any noise. WDYT?